Conflict-resolution-perf-improvements.md

```diff
@@ -0,0 +1,112 @@
+Conflict Resolution Performance Improvements
+====================
+Currently conflict resolution (CR) performance is dog slow, due to the following:
+
+1. All segments and their cache entities are checked
+    - Performance degrades rapidly as the size of the cluster increases
+
+2. Centralised CR
+    - On merge the coordinator of the majority partition is responsible for comparing all cache entities
+        - In a distributed cache, this requires the coordinator to retrieve all segments from both partitions in order
+        to compare them
+    - User triggered CR occurs on the node the request is made
+
+3. No parallelism
+    - A single segment is processed at a given time in order to avoid the centralised CR coordinator suffering a OOM exception
+
+# Proposals
+The following proposals are listed in the order of the anticipated performance benefit, with those with greatest impact
+presented first.
+
+## Maintaining a Merkle tree
+To minimise the number of entries that need to be checked during CR, we should maintain a Merkle tree of cache entry hashes
+on a per segment basis, with the root of the tree being a hash of all the entry hashes in the segment. Given the root hash
+of a segment from three different nodes, we can determine that no conflict exists simply by checking if all three values
+are equal. If any of the three hashes differ, then it's then necessary to compare the segment's entries and perform CR.
+Amazon Dynamo DB utilises this technique https://www.allthingsdistributed.com/files/amazon-dynamo-sosp2007.pdf.
+
+The simplest solution for producing such a tree, is to simple have a tree of depth 2. Where each cache entry is a leaf node
+and the root is a hash of all leaf hash values. This can be easily implemented by simply iterating over the segment container
+in order to calculate the hash of individual entries and the combined hash.
+
+TRACKED BY: [ISPN-8412] https://issues.redhat.com/browse/ISPN-8412
+
+### Hash to use
+MurmurHash3?
+
+### Calculating the entry hash
+Simply hash the `.hashcode()` returned by the `InternalCacheEntry` implementation.
+
+QUESTION: Should we include Metadata in the hash so that versions are taken into account? Are two entries with the same
+key/value but different metadata equal? Currently CR just calls `.equals` on the `CacheEntry` instances and
+`AbstractInternalCacheEntry`'s provided `equals` method only compares the key/values. Is this a bug?
```

Comment:
I'm not certain that maintaing the highest version number is always sufficient for HotRod conditional operations.

Consider the following:
1. CR is initiated when two partitions merge and `MergePolicy.PREFERRED_NON_NULL` is used.
2. Key `k` has a conflict
    - Partition 1's entry is `p1(k, v, 2)`
    - Partition 2's entry is `p2(k, v', 3)`
3. Partition `p1(k, v, 2)` is the winning entry, but version updated to 3
4. A client issues `removeWithVersion(k, 3)` after doing some processing with value `v1`. The operation succeeds, however the k/v removed from the cluster is different from that which was used by the client.

I think we need to increment an entries version after CR in which the entry with the lower version number succeeds. That way, step 4 will fail, preventing the latest value which the client has not seen from being deleted.

Reply:
I didn't mean that we should replace the version in the entry selected by the merge policy with the highest version, I agree it's a bad idea to mix the value from one entry with the version from another entry. And picking `max(versions) + 1` doesn't sound great either, because then both `removeWithVersion(k, 2)` and `removeWithVersion(k, 3)` will fail.

Since the version is a `long`, IMO properly fixing this will require adding some random "fuzz" to the version so that only clients that got the value from the "winning" partition can do successful conditional operations after a merge. Of course, if the merge policy is custom then all bets are off.

BTW we have similar problems in embedded caches, even without a merge, e.g.
https://issues.redhat.com/browse/ISPN-7638

Reply:
> Since the version is a `long`, IMO properly fixing this will require adding some random "fuzz" to the version so that only clients that got the value from the "winning" partition can do successful conditional operations after a merge. Of course, if the merge policy is custom then all bets are off.

Actually `EntryMergePolicy` returns the "winning" cache entry, so can we not just use the version associated with that entry? This is what we currently do, except we don't classify different versions as being a conflict. It would mean that the winning entry version is also determined by a custom merge policy.

Reply:
I should have been clearer, I was suggesting selecting the winning cache entry as is. 

The problem I was trying to fix is that version 2 is just a put away from version 3, so the client doing `removeWithVersion(k, 3)` just needs someone else to do a `put(k, v3)` first and then it'll succeed, removing a much newer version.